Refacto

SneakyKenny opened the pull request.

LilacOwl committed:
remove pycache

LilacOwl committed:
add Makefile

LilacOwl committed:
add utype helper function

LilacOwl committed:
updated makefile after refacto

LilacOwl committed:
add piece_type

LilacOwl committed:
add piece_position and piece_rotation

LilacOwl committed:
add piece_matrix, piece_representation and piece_helper

LilacOwl committed:
add kick_tables

LilacOwl committed:
add board

LilacOwl committed:
add basic main

LilacOwl committed:
add queue to printer

LilacOwl committed:
add line clears

LilacOwl committed:
add freeze button

LilacOwl committed:
comment out non-functionning code

LilacOwl committed:
add ability to hold piece

LilacOwl committed:
add more bindings

LilacOwl committed:
add basic board (rust)

LilacOwl committed:
move rstris to tet.rs

LilacOwl committed:
add piece movement

LilacOwl committed:
add board printer

LilacOwl committed:
add hold

LilacOwl committed:
add piece rotation

LilacOwl committed:
add line clears

LilacOwl committed:
run cargo fmt

LilacOwl committed:
WIP: add kick tables (does not compile /!\)

SanderJSA commented on tet.rs/src/main.rs:
```diff
@@ -0,0 +1,19 @@
+extern crate bit_vec;
+extern crate nanorand;
```
Not needed since 2018 Edition
```suggestion
```

SneakyKenny replied:
Got it ! Thanks.

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,382 @@
+use bit_vec::BitVec;
+use nanorand::{WyRand, RNG};
+
+use crate::tetris::board::piece::piece_matrix::{MatrixT, PieceMatrix};
+use crate::tetris::board::piece::piece_position::{PiecePosition, PositionT};
+use crate::tetris::board::piece::{
+    helper::Helper, piece_representation::PieceRepresentation, piece_rotation::PieceRotation,
+    piece_type::PieceType,
+};
```
A quick fix would be:
```suggestion
use super::piece::piece_matrix::{MatrixT, PieceMatrix};
use super::piece::piece_position::{PiecePosition, PositionT};
use super::piece::{
    helper::Helper, piece_representation::PieceRepresentation, piece_rotation::PieceRotation,
    piece_type::PieceType,
};
```
I think you are better off reexporting just what you need in piece/mod.rs instead of looking for individual parts deep in that module.

SneakyKenny replied:
Yes imports were a pain and I sort of stuck to the first solution that compiled. Thanks.

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,382 @@
+        ];
+        self.rng.shuffle(&mut bag);
+
+        let mut resulting_bag = QueueT::with_capacity(bag.len());
+        for piece in bag.iter() {
+            resulting_bag.push_back(*piece);
+        }
+
+        resulting_bag
```
```suggestion
        bag.iter().copied().collect()
```
There might be a more concise way of expressing this, but into_iter only returns a reference to PieceType for some reason.

SneakyKenny replied:
Oh thank you for that, it did seem like a horrible piece of code from my end.

SanderJSA commented on tet.rs/src/tetris/board/piece/mod.rs:
```diff
@@ -0,0 +1,7 @@
+pub mod helper;
+pub mod kick_tables;
+pub mod piece_matrix;
+pub mod piece_position;
+pub mod piece_representation;
+pub mod piece_rotation;
+pub mod piece_type;
```
You could probably keep most mods private, and only pub use what you need, otherwise the module separation doesn't make much sense.

SneakyKenny replied:
I guess what you're saying is exact.

Later on, I plan to have a pub struct in `tetris/` that's used to actually play the game. This would reference (future) helpers such as a `PieceMove` that would handle the movement of a piece, rather than trying to move around with integers that frankly don't make much sense.

For the time being, I'm keeping these modules public so that I can quickly test stuff from `main`, but they will later be made private.

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,382 @@
+    }
+
+    pub fn is_valid_move(&self, dx: PositionT, dy: PositionT, dr: PositionT) -> bool {
+        if dx != 0 && dy != 0 || dx != 0 && dr != 0 || dy != 0 && dr != 0 {
+            return false;
+        }
+
+        if dr != 0 {
+            dr.abs() <= 2
+        } else if dx != 0 {
+            dx.abs() == 1
+        } else {
+            dy.abs() == 1
+        }
```
```suggestion
        match (dx, dy, dr) {
            (_, _, dr) if dr != 0 => dr.abs() <= 2,
            (dx, _, _) if dx != 0 => dx.abs() == 1,
            (_, dy, _) if dy != 0 => dy.abs() == 1,
            _ => false,
        }

```

SneakyKenny replied:
Replacing `_`s with `0`s to make sure only one value is set.
Also reading through this snippet taugh me you can fit additionnal conditions while matching, thanks !

SanderJSA commented on tet.rs/src/tetris/board/piece/piece_matrix.rs:
```diff
@@ -0,0 +1,136 @@
+    piece_position::PositionT, piece_rotation::PieceRotation, piece_type::PieceType,
+};
+
+const IN: u16 = 0b0000111100000000;
```
Neat trick, you can add underscores in between digits which can help readability:
```suggestion
const IN: u16 = 0b0000_1111_0000_0000;
```

SneakyKenny replied:
Neat !

SanderJSA commented on tet.rs/src/tetris/board/piece/piece_matrix.rs:
```diff
@@ -0,0 +1,136 @@
+const TS: u16 = 0b000111010;
+const TW: u16 = 0b010110010;
+
+pub struct PieceMatrix {}
```
We rarely use empty structs as a way to create namespaces, creating a submodule or putting everything in its own file is preferred.

SneakyKenny replied:
Perfect ! That'll make things easier !

SanderJSA reviewed (commented):
Impressive to get this much done while still being pretty new to Rust !
I think the code would gain in readability if some aspects were written in a more functional manner.
Reorganizing how modules interact a little bit also goes a long way.

LilacOwl committed:
remove useless extern mods in main

LilacOwl committed:
cleaned board includes to make better use of modules as namespaces

LilacOwl committed:
fix possible out of bounds when displaying a large number of queue
elements

LilacOwl committed:
refacto is_valid_move logic

LilacOwl committed:
change piece modules to (again) make better use of namespaces

LilacOwl committed:
add compiling (not tested) kick tables

LilacOwl committed:
add kick table support from board's pieces rotations

LilacOwl committed:
run cargo fmt

SneakyKenny commented:
Thank you for your reply !
I've looked at each comment and they were super-duper-useful !

I'm currently missings tests, even the commits pushed just now did not get test what-so-ever.

I'm looking to implement some keyboard I/O that will help properly test (functional tests) everything.

LilacOwl committed:
add piece color

LilacOwl committed:
add termion for I/O

LilacOwl committed:
add piece colors

LilacOwl committed:
make get_cell and set_cell functions return Results instead

Known issue: clear_completed_lines makes lines disappear when they
should not, so it was commented out.

LilacOwl committed:
add todo

LilacOwl committed:
fix line clears

Also run formatter (and added to vimrc so it doesn't happen again)

LilacOwl committed:
add complete piece color support

LilacOwl committed:
reset cursor position before printing board

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,497 @@
+
+                // FIXME: properly check both validity of position and
+                // board state at this position
+                if x + px >= BOARD_WIDTH
+                    || y + py >= BOARD_HEIGHT * 2
+                    || self.get_cell(x + px, y + py).is_none()
+                    || self.get_cell(x + px, y + py).unwrap()
```
```suggestion
                if self.get_cell(x + px, y + py) != Some(false)
```

SneakyKenny replied:
I may have forgotten to clean old cold on this one...

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,497 @@
+    }
+
+    fn clear_completed_lines(&mut self) {
+        let mut completed_lines: [bool; BOARD_HEIGHT as usize * 2] =
+            [false; BOARD_HEIGHT as usize * 2];
+
+        for y in 0..(BOARD_HEIGHT * 2) {
+            if self.is_line_completed(y) {
+                completed_lines[y as usize] = true;
+            }
+        }
```
```suggestion
             let completed_lines: Vec<bool> = (0..(BOARD_HEIGHT * 2))
            .map(|y| self.is_line_completed(y))
            .collect();
```

SneakyKenny replied:
Looks much better ! 👍

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,497 @@
+        y: piece_position::PositionT,
+        val: OptionalTypeT,
+    ) {
+        match Board::xy_to_index(x, y) {
+            Ok(index) => self.piece_types[index] = val,
+            Err(_) => {}
```
```suggestion
        if let Ok(index) = Board::xy_to_index(x, y) {
            self.piece_types[index] = val;
```

SneakyKenny replied:
Much better. Thanks.

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,497 @@
+        y: piece_position::PositionT,
+        val: bool,
+    ) {
+        match Board::xy_to_index(x, y) {
+            Ok(index) => self.board.set(index, val),
+            Err(_) => {}
```
```suggestion
        if let Ok(index) = Board::xy_to_index(x, y) {
            self.board.set(index, val);
```

SneakyKenny replied:
I'm still a bit confused by this line, I'll have to think when reading through it later !
Still applying.

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,497 @@
+        true
+    }
+
+    fn xy_to_index(
+        x: piece_position::PositionT,
+        y: piece_position::PositionT,
+    ) -> Result<usize, &'static str> {
+        if x < 0 || y < 0 || x >= BOARD_WIDTH || y >= BOARD_HEIGHT * 2 {
+            Err("Coordinate is outside the board.")
+        } else {
+            Ok((y * BOARD_WIDTH + x) as usize)
+        }
+    }
```
```suggestion
    fn xy_to_index(x: piece_position::PositionT, y: piece_position::PositionT) -> Option<usize> {
        if (0..BOARD_WIDTH).contains(&x) && (0..BOARD_HEIGHT * 2).contains(&y) {
            Some((y * BOARD_WIDTH + x) as usize)
        } else {
            None
        }
    }

```

SneakyKenny replied:
Funky !

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,497 @@
+        x: piece_position::PositionT,
+        y: piece_position::PositionT,
+    ) -> OptionalTypeT {
+        match Board::xy_to_index(x, y) {
+            Ok(index) => self.piece_types[index],
+            Err(_) => None,
+        }
```
```suggestion
        Board::xy_to_index(x, y).and_then(|index| self.piece_types[index])
```

SneakyKenny replied:
Way better.

SanderJSA commented on tet.rs/src/tetris/board/my_board.rs:
```diff
@@ -0,0 +1,497 @@
+        x: piece_position::PositionT,
+        y: piece_position::PositionT,
+    ) -> Option<bool> {
+        match Board::xy_to_index(x, y) {
+            Ok(index) => Some(self.board[index]),
+            Err(_) => None,
+        }
```
```suggestion
        Board::xy_to_index(x, y).map(|index| self.board[index])
```

SneakyKenny replied:
Neat.

SanderJSA commented on tet.rs/src/tetris/mod.rs:
```diff
@@ -0,0 +1,2 @@
+pub mod board;
+pub mod my_tetris;
```
Tetris module should probably be merged with main.rs

SneakyKenny replied:
Refacto is a WIP !

SanderJSA reviewed (commented):
Some aspects can be made more functional, but great overall

LilacOwl committed:
apply suggestions for code readability
